contracts/javatest/src/test/java/net/consensys/gpact/functioncall/gpact/FailedGpactCallTreeTest.java

```diff
@@ -0,0 +1,298 @@
+    BigInteger callDepthLeaf = BigInteger.TWO;
+
+    ArrayList<CallExecutionTree> intermediateCalls = new ArrayList<>();
+    ArrayList<CallExecutionTree> rootCalls = new ArrayList<>();
+    CallExecutionTree leafSeg =
+        new CallExecutionTree(
+            bc3.bcId,
+            leafBlockchain.contractAddress(),
+            intermediateBlockchain.getAbi(totalCallDepth, callDepthLeaf, failAtCallDepth));
+    intermediateCalls.add(leafSeg);
+    CallExecutionTree intermediateSeg =
+        new CallExecutionTree(
+            bc2.bcId,
+            intermediateBlockchain.contractAddress(),
+            intermediateBlockchain.getAbi(totalCallDepth, callDepthIntermediate, failAtCallDepth),
+            intermediateCalls);
+    rootCalls.add(intermediateSeg);
+    CallExecutionTree rootCall =
+        new CallExecutionTree(
+            root.bcId,
+            rootBlockchain.contractAddress(),
+            rootBlockchain.getAbi(totalCallDepth, callDepthRoot, failAtCallDepth),
+            rootCalls);
+
+    CrosschainCallResult result =
+        crossControlManagerGroup.executeCrosschainCall(
+            exampleManager.getExecutionEngine(), rootCall, 300);
+
+    if (result.isSuccessful()) {
+      throw new Exception("FailLeafSegment was unexpectedly successful");
+    }
+  }
+
+  private void failRoot() throws Exception {
+    LOG.info("****** Fail Root *******");
+
+    BigInteger totalCallDepth = BigInteger.TWO;
+    BigInteger failAtCallDepth =
+        BigInteger.ZERO; // Fail after the total call depth, hence don't fail.
+    BigInteger callDepthRoot = BigInteger.ZERO;
+    BigInteger callDepthIntermediate = BigInteger.ONE;
+    BigInteger callDepthLeaf = BigInteger.TWO;
+
+    ArrayList<CallExecutionTree> intermediateCalls = new ArrayList<>();
+    ArrayList<CallExecutionTree> rootCalls = new ArrayList<>();
+    CallExecutionTree leafSeg =
+        new CallExecutionTree(
+            bc3.bcId,
+            leafBlockchain.contractAddress(),
+            intermediateBlockchain.getAbi(totalCallDepth, callDepthLeaf, failAtCallDepth));
+    intermediateCalls.add(leafSeg);
+    CallExecutionTree intermediateSeg =
+        new CallExecutionTree(
+            bc2.bcId,
+            intermediateBlockchain.contractAddress(),
+            intermediateBlockchain.getAbi(totalCallDepth, callDepthIntermediate, failAtCallDepth),
+            intermediateCalls);
+    rootCalls.add(intermediateSeg);
+    CallExecutionTree rootCall =
+        new CallExecutionTree(
+            root.bcId,
+            rootBlockchain.contractAddress(),
+            rootBlockchain.getAbi(totalCallDepth, callDepthRoot, failAtCallDepth),
+            rootCalls);
+
+    CrosschainCallResult result =
+        crossControlManagerGroup.executeCrosschainCall(
+            exampleManager.getExecutionEngine(), rootCall, 300);
+
+    if (result.isSuccessful()) {
+      throw new Exception("FailLeafSegment was unexpectedly successful");
```

Comment:
these lines are largely duplicated across the functions, is it possible to refactor out the commonality?

Reply:
I have abstracted all of this into a single function so all four sub-tests use the one function / piece of common code.